pkg/forward/forward.go

```diff
@@ -385,45 +388,49 @@ func (isdf *InterStepDataForward) forwardAChunk(ctx context.Context) {
+		if publisher, ok := isdf.wmPublishers[toVertexName]; ok {
 			for index, offsets := range toVertexBufferOffsets {
 				if isdf.opts.vertexType == dfv1.VertexTypeSource || isdf.opts.vertexType == dfv1.VertexTypeMapUDF ||
 					isdf.opts.vertexType == dfv1.VertexTypeReduceUDF {
 					if len(offsets) > 0 {
 						publisher.PublishWatermark(processorWM, offsets[len(offsets)-1], int32(index))
-						activeWatermarkBuffers[toVertexName] = true
+						activeWatermarkBuffers[toVertexName][index] = true
 						// reset because the toBuffer is no longer idling
 						isdf.idleManager.Reset(toVertexName)
 					}
 					// This (len(offsets) == 0) happens at conditional forwarding, there's no data written to the buffer
 				} else { // For Sink vertex, and it does not care about the offset during watermark publishing
 					publisher.PublishWatermark(processorWM, nil, int32(index))
-					activeWatermarkBuffers[toVertexName] = true
+					activeWatermarkBuffers[toVertexName][index] = true
 					// reset because the toBuffer is no longer idling
 					isdf.idleManager.Reset(toVertexName)
 				}
 			}
 		}
 	}
 	// TODO(multi-partition): handle idle watermark publishing for multi partitioned buffer
-	if len(dataMessages) > 0 && len(activeWatermarkBuffers) < len(isdf.publishWatermark) {
-		// - condition1 "len(dataMessages) > 0" :
-		//   Meaning, we do have some data messages, but not all out buffers get written.
-		//   Tt could be all data messages are dropped, or conditional forwarding to part of the out buffers.
-		//   If we don't have this condition check, when dataMessages is zero but ctrlMessages > 0, we will
-		//   wrongly publish an idle watermark without the ctrl message and the ctrl message tracking map.
-		// - condition 2 "len(activeWatermarkBuffers) < len(isdf.publishWatermark)" :
-		//   send idle watermark only if we have idle out buffers
-		for bufferName := range isdf.publishWatermark {
-			if !activeWatermarkBuffers[bufferName] {
-				// use the watermark of the current read batch for the idle watermark
-				// same as read len==0 because there's no event published to the buffer
-				if p, ok := isdf.publishWatermark[bufferName]; ok {
-					idlehandler.PublishIdleWatermark(ctx, isdf.toBuffers[bufferName][0], p, isdf.idleManager, 0, isdf.opts.logger, isdf.opts.vertexType, processorWM)
+	// - condition1 "len(dataMessages) > 0" :
+	//   Meaning, we do have some data messages, but we may not have written to all out buffers or its partitions.
+	//   It could be all data messages are dropped, or conditional forwarding to part of the out buffers.
+	//   If we don't have this condition check, when dataMessages is zero but ctrlMessages > 0, we will
+	//   wrongly publish an idle watermark without the ctrl message and the ctrl message tracking map.
+	// - condition 2 "len(activeWatermarkBuffers) < len(isdf.wmPublishers)" :
+	//   send idle watermark only if we have idle out buffers
+	if len(dataMessages) > 0 {
```

Comment:
We've agreed that, in the condition where the readMessage contains only control messages, we don't execute extra steps: because if the idling situation continues, the next `readMessages, err := isdf.fromBuffer.Read(ctx, isdf.opts.readBatchSize)` will have `len(readMessages)==0`, and we eventually will handle the idling.

Reply:
I think in reduce, we return early if len(dataMessages) is zero
so removing the `p.result>0` is the situation where we do have data messages, but after applying the reduce UDF, we don’t have any valid outputs, all messages are dropped, and we send idle watermark in this situation...

so the change looks good to me
no need to add the check